contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -320,6 +348,583 @@ babelfish_concat_wrapper(PG_FUNCTION_ARGS)
+
+		if (timestamp2tm(timestamp, &tz1, tm, &fsec1, NULL, NULL) != 0)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
+				errmsg("timestamp out of range")));
+		}
+
+		year = tm->tm_year;
+		month = tm->tm_mon;
+		day = tm->tm_mday;
+
+		if (strcmp(field, "year") == 0)
+		{
+			return tm->tm_year;
+		}
+		else if (strcmp(field, "quarter") == 0)
+		{
+			return (int)ceil((float)tm->tm_mon / 3.0);
+		}
+		else if (strcmp(field, "month") == 0)
+		{
+			return tm->tm_mon;
+		}
+		else if (strcmp(field, "day") == 0)
+		{
+			return tm->tm_mday;
+		}
+		else if (strcmp(field, "hour") == 0)
+		{
+			return tm->tm_hour;
+		}
+		else if (strcmp(field, "minute") == 0)
+		{
+			return tm->tm_min;
+		}
+		else if (strcmp(field, "second") == 0)
+		{
+			return tm->tm_sec;
+		}
+		else if (strcmp(field, "doy") == 0)		/* day-of-year of the date */
+		{
+			return (date2j(tm->tm_year, tm->tm_mon, tm->tm_mday)
+							 - date2j(tm->tm_year, 1, 1) + 1);
+		}
+		else if (strcmp(field, "dow") == 0)		/* day-of-week of the date */
+		{
+			/* dow calculated using Zeller's Congruence */
+			if (tm->tm_mon < 3)
+			{
+				month += MONTHS_PER_YEAR;
+				year -= 1;
+			}
+
+			/* 
+			 * Zeller’s congruence is an algorithm devised by Christian Zeller to calculate
+			 * the day of the week for any calendar date. 
+			 * https://www.themathdoctors.org/zellers-rule-what-day-of-the-week-is-it/
+			 * Below is the equation
+			 */
+			res = (day + 2*month + ((3*(month+1))/(5)) + year + year/4 - year/100 + year/400 + 2) % 7;
+			
+			/* Adjusting the dow accourding to the datefirst guc */
+			return ((res) + 7 - pltsql_datefirst)%7 == 0 ?
+						7 : ((res) + 7 - pltsql_datefirst)%7;
+
+		}
+		else if (strcasecmp(field , "tsql_week") == 0)		/* week number of the year  */
+		{
+			first_day = date2j(tm->tm_year, 1, 1) - UNIX_EPOCH_JDATE; /* returns number of days since 1/1/1970 to 1/1/tm_year */
+			/* convert this first day of tm_year to timestamp into tsql_first_day */
+			tsql_first_day = (Timestamp) (first_day - (POSTGRES_EPOCH_JDATE - UNIX_EPOCH_JDATE)) * USECS_PER_DAY;
+			first_week_end = 8 - datepart_internal("dow", tsql_first_day, 0, false);
+
+			day_of_year = datepart_internal("doy",timestamp,0, false);
+
+			if(day_of_year <= first_week_end)
+			{
+				return 1;		/* day of year is less than first_week_end means its a first week */
+			}
+			else
+			{
+				return (2+(day_of_year - first_week_end - 1) / 7);
+			}
+		}
+		else if(strcasecmp(field , "week") == 0)
+		{
+			return date2isoweek(tm->tm_year, tm->tm_mon, tm->tm_mday);
+		}
+		else if(strcasecmp(field , "millisecond") == 0)
+		{
+			return (fsec1) / 1000;
+		}
+		else if(strcasecmp(field , "microsecond") == 0)
+		{
+			return (fsec1);
+		}
+		else if(strcasecmp(field , "nanosecond") == 0)
+		{
+			return (fsec1) * 1000;
+		}
+		else if(strcasecmp(field , "tzoffset") == 0 && !general_integer_datatype)
```

Comment:
SQL Server behaviour when general_integer_datatype = true is that it throws specific error. Lets support that